### DISCUSSION TITLE
Can we further narrow down the default accessible locations in Privacy Mode?

### DISCUSSION THREAD

**Discussion by gexgd0419** [sandboxie-plus/Sandboxie#2371] [General]
To allow programs to work properly, some system locations are still accessible in Privacy Mode.

I'm not sure what exactly are those locations, but at least these locations seem accessible by default:

- `%SystemRoot%`
- `%ProgramFiles%`
- `%SystemDrive%\Program Files (x86)` (sorry but I didn't find a variable that can represent the x86 version of `%ProgramFiles%`)
- `HKEY_LOCAL_MACHINE`

which are mostly system locations, instead of user locations that may contain sensitive data.

But, in my opinion, this range can still be narrowed down.

Considering that `%ProgramFiles%` is the default location for software installation, and some programs also put global configuration data inside `%ProgramFiles%` and `HKEY_LOCAL_MACHINE\SOFTWARE`, allowing sandboxed programs to access those locations could leak some information about, for example, what software the user has installed on the system, which I think is also a part of the user's privacy.

As most items created by other applications in `%ProgramFiles%` and `HKEY_LOCAL_MACHINE\SOFTWARE` is not necessary for sandboxed programs, their ranges can be narrowed to their sub-items that come with Windows.

So I turned off Privacy Mode, turned on `UseRuleSpecificity`, and tried to use custom rules to define a more-enhanced "Privacy Mode".

Below are the rules I've come up with, as for now. Improvements are welcome.

<details>
<summary>Sandboxie INI section</summary>

``` Ini
UseRuleSpecificity=y

# File system rules

# Default behavior: write only
WriteFilePath=*

# Allow access to Sandboxie programs
ReadFilePath=C:\Program Files\Sandboxie-Plus

# Allowed Program Files sub-folders
NormalFilePath=%ProgramFiles%\Common Files\Microsoft Shared
NormalFilePath=%ProgramFiles%\Common Files\Services
NormalFilePath=%ProgramFiles%\Common Files\System
NormalFilePath=%SystemDrive%\Program Files (x86)\Common Files\Microsoft Shared
NormalFilePath=%SystemDrive%\Program Files (x86)\Common Files\Services
NormalFilePath=%SystemDrive%\Program Files (x86)\Common Files\System
NormalFilePath=%ProgramFiles%\Internet Explorer
NormalFilePath=%ProgramFiles%\Windows Defender
NormalFilePath=%ProgramFiles%\Windows Mail
NormalFilePath=%ProgramFiles%\Windows Media Player
NormalFilePath=%ProgramFiles%\Windows Multimedia Platform
NormalFilePath=%ProgramFiles%\Windows NT
NormalFilePath=%ProgramFiles%\Windows Photo Viewer
NormalFilePath=%ProgramFiles%\Windows Portable Devices
NormalFilePath=%ProgramFiles%\Windows Sidebar
NormalFilePath=%ProgramFiles%\WindowsPowershell
NormalFilePath=%SystemDrive%\Program Files (x86)\Internet Explorer
NormalFilePath=%SystemDrive%\Program Files (x86)\Windows Defender
NormalFilePath=%SystemDrive%\Program Files (x86)\Windows Mail
NormalFilePath=%SystemDrive%\Program Files (x86)\Windows Media Player
NormalFilePath=%SystemDrive%\Program Files (x86)\Windows Multimedia Platform
NormalFilePath=%SystemDrive%\Program Files (x86)\Windows NT
NormalFilePath=%SystemDrive%\Program Files (x86)\Windows Photo Viewer
NormalFilePath=%SystemDrive%\Program Files (x86)\Windows Portable Devices
NormalFilePath=%SystemDrive%\Program Files (x86)\Windows Sidebar
NormalFilePath=%SystemDrive%\Program Files (x86)\WindowsPowershell
NormalFilePath=%ProgramFiles%\Windows Defender Advanced Threat Protection
NormalFilePath=%ProgramFiles%\Windows Security
NormalFilePath=%ProgramFiles%\WindowsApps
NormalFilePath=%SystemDrive%\Program Files (x86)\Microsoft
NormalFilePath=%SystemDrive%\Program Files (x86)\Microsoft.NET

# %SystemRoot%
NormalFilePath=%SystemRoot%
WriteFilePath=%SystemRoot%\Tasks
WriteFilePath=%SystemRoot%\System32\Tasks
WriteFilePath=%SystemRoot%\SysWOW64\Tasks
WriteFilePath=%SystemRoot%\Logs
WriteFilePath=%SystemRoot%\System32\Logs
WriteFilePath=%SystemRoot%\System32\LogFiles
WriteFilePath=%SystemRoot%\SysWOW64\LogFiles
WriteFilePath=%SystemRoot%\System32\winevt\Logs
WriteFilePath=%SystemRoot%\Minidump
WriteFilePath=%SystemRoot%\Temp

# Registry keys rules

# Default behavior: write only
WriteKeyPath=*

# Allow access to subkeys of HKLM, except SOFTWARE
NormalKeyPath=HKEY_LOCAL_MACHINE
WriteKeyPath=HKEY_LOCAL_MACHINE\SOFTWARE

# Blocked subkeys of SOFTWARE
WriteKeyPath=HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Run
WriteKeyPath=HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\RunOnce
WriteKeyPath=HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall
WriteKeyPath=HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\RADAR\HeapLeakDetection\DiagnosedApplications
WriteKeyPath=HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Schedule
WriteKeyPath=HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Microsoft\Windows\CurrentVersion\Run
WriteKeyPath=HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Microsoft\Windows\CurrentVersion\RunOnce
WriteKeyPath=HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Microsoft\Windows\CurrentVersion\Uninstall
WriteKeyPath=HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Microsoft\RADAR\HeapLeakDetection\DiagnosedApplications
WriteKeyPath=HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Microsoft\Windows NT\CurrentVersion\Schedule
WriteKeyPath=HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Explorer\StartupApproved

# Allowed subkeys of SOFTWARE
NormalKeyPath=HKEY_LOCAL_MACHINE\SOFTWARE\Classes
NormalKeyPath=HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft
NormalKeyPath=HKEY_LOCAL_MACHINE\SOFTWARE\ODBC
NormalKeyPath=HKEY_LOCAL_MACHINE\SOFTWARE\Policies
NormalKeyPath=HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Classes
NormalKeyPath=HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Microsoft
NormalKeyPath=HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\ODBC
NormalKeyPath=HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\Policies
NormalKeyPath=HKEY_LOCAL_MACHINE\SOFTWARE\DefaultUserEnvironment
```

</details>

Its range is still larger than necessary, though.

Compared to the real Privacy Mode, using a normal sandbox with the rules above will cause a problem.
If the sandbox is empty, during its initialization, it will fail to initialize the registry, and it won't create standard shell folders.
If you first turn on Privacy Mode to run a program in the sandbox, then turn off Privacy Mode and apply the rules above, it will work as intended.

My goal is to leak as little information to sandboxed programs as possible (and the "ultimate goal" is to make sandboxed programs unable to distinguish the system they're running on from a freshly installed one). Of course, hardware-related information cannot be protected this way, and using a virtual machine may be a better way, but I still hope Sandboxie's Privacy Mode to be more secure, while still being user-friendly.

Some may argue that this will make non-system apps installed in `%ProgramFiles%` not be able to run in the sandbox. But if the program is not in `%ProgramFiles%`, for example, it gets installed in disk D: or is a portable exe in Downloads folder, it won't be able to run anyway. Users still have to add a resource access rule for it. (Also it would be better to prompt users to modify resource access rules when the program to run in the sandbox is denied access according to the rules, instead of just saying "access is denied" or "file not found")

**Comment by bastik-1001** [comment]
Thank you for bringing this up and providing some improvement, as well as hopefully valuable feedback.

Considering how long Sandboxie has been around and how many things got improved recently, we can be thankful that David (and other contributors) is putting so much work into it. Given that the Privacy Mode is quite new, there are probably still things that can be improved, as you point out.

I am even sure that this can be done in a user-friendly way, but there are probably things that the user needs to configure in some way to make his use-case work.

**Comment by DavidXanatos** [comment]
This would definitely be an improvement, I just went with the easy list for start to avoid too many incompatibilities.
I could add a function to check if a template with a predefined name exists and if so use it instead of the built in list.
Also you can overwrite the built in list by specifying the same paths but as write only paths...
https://github.com/sandboxie-plus/Sandboxie/blob/44fdbba1995c1bb1633cc6232af81958f441aa9d/Sandboxie/core/drv/file.c#L583
https://github.com/sandboxie-plus/Sandboxie/blob/44fdbba1995c1bb1633cc6232af81958f441aa9d/Sandboxie/core/drv/key.c#L244

**Comment by isaak654** [comment]
@DavidXanatos 
See https://github.com/sandboxie-plus/Sandboxie/issues/2402
